src/inmanta/agent/forking_executor.py

```diff
@@ -276,10 +327,36 @@ def init_once(cls) -> None:
             # already set
             pass
 
-    async def create_executor(self, venv: executor.ExecutorVirtualEnvironment, executor_id: executor.ExecutorId) -> MPExecutor:
+    async def get_executor(
+        self, agent_name: str, agent_uri: str, code: typing.Collection[executor.ResourceInstallSpec]
+    ) -> MPExecutor:
+        """
+        Retrieves an Executor based on the agent name and blueprint.
+        If an Executor does not exist for the given configuration, a new one is created.
+
+        :param agent_name: The name of the agent for which an Executor is being retrieved or created.
+        :param blueprint: The ExecutorBlueprint defining the configuration for the Executor.
+        :return: An Executor instance
+        """
+        blueprint = executor.ExecutorBlueprint.from_specs(code)
+        executor_id = executor.ExecutorId(agent_name, agent_uri, blueprint)
+        if executor_id in self.executor_map:
+            return self.executor_map[executor_id]
+        # Acquire a lock based on the blueprint's hash
+        # We don't care about URI here
+        async with self._locks.get(executor_id.identity()):
+            if executor_id in self.executor_map:
+                return self.executor_map[executor_id]
+            my_executor = await self.create_executor(executor_id)
+            self.__add_executor(executor_id, my_executor)
+            return my_executor
+
+    async def create_executor(self, executor_id: executor.ExecutorId) -> MPExecutor:
         # entry point from parent class
-        executor = await self.make_child_and_connect(executor_id.agent_name)
-        storage_for_blueprint = os.path.join(self.storage_folder, "code", executor_id.blueprint.generate_blueprint_hash())
+        env_blueprint = executor_id.blueprint.to_env_blueprint()
+        venv = await self.environment_manager.get_environment(env_blueprint, self.thread_pool)
+        executor = await self.make_child_and_connect(executor_id, venv)
+        storage_for_blueprint = os.path.join(self.storage_folder, "code", executor_id.blueprint.blueprint_hash())
```

Comment:
Why not use `executor_id.identity()` instead? It would be more representative to a human inspecting the file system.

Reply:
it would not be able to share this folder over agent instances in that case.

Reply:
it would make the cache miss more often on very similar agents,...